File: goon_http/handlers/sgg_discord_bot_handler.py
import asyncio
import hashlib
import hmac
import logging

from flask import Blueprint, request, abort

from discord_bot.discord_bot import get_current_bot
from discord_bot.webhook_reactions.release_event import post_new_release
from discord_bot.webhook_reactions.issue_event import handle_issue_event

log = logging.getLogger(__name__)

sgg_discord_bot_bp = Blueprint("sgg_discord_bot", __name__)

SECRET = "kevinb"


@sgg_discord_bot_bp.route("/api/v1/webhooks/sggDiscordBot", methods=["POST"])
def handle_supergoon_games_discord_bot():
    body = request.get_data()

    signature = request.headers.get("X-Hub-Signature", "")
    if not verify_signature(signature, body):
        log.warning("Invalid signature")
        abort(401)

    event_type = request.headers.get("X-GitHub-Event", "")
    bot = get_current_bot()
    loop = bot.loop

    if event_type == "release":
        payload = request.get_json()
        if payload.get("action") == "published":
            release = payload["release"]
            repository = payload["repository"]
            asyncio.run_coroutine_threadsafe(
                post_new_release(
                    repository["html_url"],
                    release["html_url"],
                    release["name"],
                    release["body"],
                    release["tag_name"],
                ),
                loop,
            ).result(timeout=10)
    elif event_type == "issues":
        payload = request.get_json()
        asyncio.run_coroutine_threadsafe(handle_issue_event(payload), loop).result(timeout=10)

    return "", 200


def verify_signature(signature: str, payload: bytes) -> bool:
    expected_mac = calculate_mac(payload)
    return hmac.compare_digest(signature, expected_mac)


def calculate_mac(payload: bytes) -> str:
    mac = hmac.new(SECRET.encode(), payload, hashlib.sha1)
    return "sha1=" + mac.hexdigest()

File: go/goon_http/handlers/sgg_discord_bot_handler.go
package handlers

import (
	"crypto/hmac"
	"crypto/sha1"
	"encoding/hex"
	"encoding/json"
	"io"
	"log"
	"net/http"

	"sgg-discord-bot/discord_bot/webhook_reactions"
)

const secret = "kevinb"

func RegisterHandlers(mux *http.ServeMux) {
	mux.HandleFunc("POST /api/v1/webhooks/sggDiscordBot", handleSupergoonGamesDiscordBot)
}

func handleSupergoonGamesDiscordBot(w http.ResponseWriter, r *http.Request) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "Bad request", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	signature := r.Header.Get("X-Hub-Signature")
	if !verifySignature(signature, body) {
		log.Println("Invalid signature")
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	eventType := r.Header.Get("X-GitHub-Event")

	switch eventType {
	case "release":
		var payload map[string]interface{}
		if err := json.Unmarshal(body, &payload); err != nil {
			break
		}
		if action, _ := payload["action"].(string); action == "published" {
			release, _ := payload["release"].(map[string]interface{})
			repository, _ := payload["repository"].(map[string]interface{})
			webhook_reactions.PostNewRelease(
				getString(repository, "html_url"),
				getString(release, "html_url"),
				getString(release, "name"),
				getString(release, "body"),
				getString(release, "tag_name"),
			)
		}
	case "issues":
		var payload map[string]interface{}
		if err := json.Unmarshal(body, &payload); err != nil {
			break
		}
		webhook_reactions.HandleIssueEvent(payload)
	}

	w.WriteHeader(http.StatusOK)
}

func verifySignature(signature string, payload []byte) bool {
	expectedMAC := calculateMAC(payload)
	return hmac.Equal([]byte(signature), []byte(expectedMAC))
}

func calculateMAC(payload []byte) string {
	mac := hmac.New(sha1.New, []byte(secret))
	mac.Write(payload)
	return "sha1=" + hex.EncodeToString(mac.Sum(nil))
}

func getString(m map[string]interface{}, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}
